Better UX error for migrations

uilianries opened the pull request:
I didn't find specific test file only related to Migrator or ClientMigrator, so I created an unit test to validate the current error message.

This change should affect more tests, as now `save` raises an error. Let's see ...

Changelog: Fix: Improve permission error message when migrating cache folder.
Docs: Omit

Related to #7948

/cc @ngrodzitski @memsharded 

- [ ] Refer to the issue that supports this Pull Request.
- [ ] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [ ] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [ ] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

uilianries committed:
Better UX error for migrations

Signed-off-by: Uilian Ries <uilianries@gmail.com>

memsharded commented on conans/util/files.py:
```diff
@@ -188,6 +188,13 @@ def save(path, content, only_if_modified=False, encoding="utf-8"):
+            raise OSError("Could not create the directory '{}' due to lack of permission."
+                          " Please, check your user write permission.".format(path))
+        elif error.errno == errno.EEXIST:
+            raise OSError("Could not create the directory '{}' because it already exists."
```
If it exists it is OK, no need to raise!

uilianries replied:
Yes, after seeing tons of test errors on my local environment, I agree :joy:

uilianries committed:
File already exists is not an error

Signed-off-by: Uilian Ries <uilianries@gmail.com>

memsharded commented on conans/util/files.py:
```diff
@@ -188,6 +188,10 @@ def save(path, content, only_if_modified=False, encoding="utf-8"):
+    except OSError as error:
+        if error.errno == errno.EACCES:
+            raise OSError("Could not create the directory '{}' due to lack of permission."
+                          " Please, check your user write permission.".format(path))
```
Directory name is ``os.path.dirname(path)``, not ``path``

uilianries replied:
Done

uilianries committed:
Print dir path name for makedirs error

Signed-off-by: Uilian Ries <uilianries@gmail.com>

uilianries committed:
Fix cache attribute

Signed-off-by: Uilian Ries <uilianries@gmail.com>

uilianries committed:
Run migration test

Signed-off-by: Uilian Ries <uilianries@gmail.com>

uilianries committed:
Fix expected string

Signed-off-by: Uilian Ries <uilianries@gmail.com>

memsharded commented on conans/client/migrations.py:
```diff
@@ -95,7 +95,7 @@ def _make_migrations(self, old_version):
 
         if old_version < Version("1.31.0"):
-            migrate_tgz_location(self.cache, self.out)
+            migrate_tgz_location(cache, self.out)
```
This is weird, it is already fixed in develop

uilianries replied:
lemme force a pull here

uilianries replied:
Okay, fixed after doing a new pull.

uilianries committed:
Merge branch 'develop' of https://github.com/conan-io/conan into ux/home-permission

memsharded commented on conans/util/files.py:
```diff
@@ -189,6 +189,10 @@ def save(path, content, only_if_modified=False, encoding="utf-8"):
     try:
         os.makedirs(os.path.dirname(path))
+    except OSError as error:
+        if error.errno == errno.EACCES:
```
I think this has some potential to break something. The folder might exist, but created by another user, and instead of getting a EEXIST error, get an EACCES error? Who knows.

I think I would make sense to check the existence of the folder, something like:

```python
folder = os.path.dirname(path)
if not os.path.isdir(folder):
     try:
         os.makedirs(folder)
     except Exception as e:
         raise OSError("The folder {} does not exist and could not be created: {}".format(folder, str(e)))
```

uilianries replied:
On Linux I'm sure it will result in a Permission error (errno.EACCES) if I try to re-create a folder which already exists, but created by another user. Anyway, I think don't harm using your logic, because I can say nothing about OSX and Windows behavior for this case.

uilianries replied:
Done.

memsharded commented on conans/migrations.py:
```diff
@@ -32,6 +32,8 @@ def _make_migrations(self, old_version):
     def _update_version_file(self):
         try:
             save(self.file_version_path, str(self.current_version))
+        except OSError as error:
```
Better just capture ``except Exception as error`` and include the ``str(error)`` message in the "Can't write version file in...", as it would be more complete and informative

uilianries replied:
Yes, as utils.file.save raises an informative message, my concern was about duplicated message. But in terms of code, it's too generic here, we can't deduce the problem without checking save() first.

uilianries replied:
Done.

uilianries committed:
Improve migration error message

Signed-off-by: Uilian Ries <uilianries@gmail.com>

uilianries committed:
Filter error message for utils.file.save

Signed-off-by: Uilian Ries <uilianries@gmail.com>

uilianries commented on conans/util/files.py:
```diff
@@ -187,10 +187,16 @@ def save(path, content, only_if_modified=False, encoding="utf-8"):
+        try:
+            os.makedirs(os.path.dirname(path))
+        except OSError as error:
+            if "permission denied" in error.strerror.lower():
```
I don't like it, but it's required because of test_get_gunzip. Maybe we should use errno.EACCES instead

memsharded replied:
This indeed reads ugly and shouldn't be necessary. Maybe the test should be reviewed too.
It should be simpler, we already checked that the folder didn't exist. So, except some weird concurrency issue, this makedirs should succeed and create the folder. Probably just checking errno.EEXIST, just to prevent any concurrency issue, and any other makedir exception should just be reraised (with an improved error message if necessary)

uilianries replied:
Sounds a good solution

uilianries replied:
The test `test_get_gunzip` raises `errno.ENOENT`, or "No such file or directory"

memsharded commented on conans/util/files.py:
```diff
@@ -187,10 +187,16 @@ def save(path, content, only_if_modified=False, encoding="utf-8"):
-        pass
+    dir_path = os.path.dirname(path)
+    if not os.path.isdir(dir_path):
+        errmsg = "The folder {} does not exist and could not be created ({})."
```
Only create the error message when it is going to be launched

uilianries replied:
Done.

memsharded commented on conans/util/files.py:
```diff
@@ -187,10 +187,16 @@ def save(path, content, only_if_modified=False, encoding="utf-8"):
+    if not os.path.isdir(dir_path):
+        errmsg = "The folder {} does not exist and could not be created ({})."
+        try:
+            os.makedirs(os.path.dirname(path))
```
Use ``dir_path`` here.

uilianries replied:
Done.

uilianries committed:
Exclude errno.ENOENT from file errors

Signed-off-by: Uilian Ries <uilianries@gmail.com>

uilianries committed:
Could not apply chmod on Windows

There is a note on chmod docs:
https://docs.python.org/2.7/library/os.html#os.chmod

"Although Windows supports chmod(), you can only set the file’s
read-only flag with it (via the stat.S_IWRITE and stat.S_IREAD constants
or a corresponding integer value). All other bits are ignored."

However, it works for files, not for folder.

Signed-off-by: Uilian Ries <uilianries@gmail.com>

uilianries commented:
I skipped Windows test because os.chmod has no effect under Windows folders, I would need other libraries, like [pywin32](https://sourceforge.net/projects/pywin32) or [oschmod](https://github.com/YakDriver/oschmod/blob/main/oschmod/__init__.py). 

The documentation is clear about Windows support: https://docs.python.org/2.7/library/os.html#os.chmod

> Although Windows supports chmod(), you can only set the file’s read-only flag with it (via the stat.S_IWRITE and stat.S_IREAD constants or a corresponding integer value). All other bits are ignored.

It works for file only ... 

For some reason, if I inject `os.stat()` before calling `os.makedirs()` it works, but I won't change `save()` only because a single test.

I also tried subprocess call with `attrib +R dir_path /D` but had no effect.
